src/errorpage.cc

```diff
@@ -810,67 +812,59 @@ ErrorState::~ErrorState()
 int
 ErrorState::Dump(MemBuf * mb)
 {
-    MemBuf str;
-    char ntoabuf[MAX_IPSTRLEN];
+    PackableStream out(*mb);
+    const auto &encoding = CharacterSet::RFC3986_UNRESERVED();
+
+    out << "?subject=" <<
+        AnyP::Uri::Encode(SBuf("CacheErrorInfo - "),encoding) <<
+        AnyP::Uri::Encode(SBuf(errorPageName(type)), encoding);
+
+    SBufStream body;
+    body << "CacheHost: " << getMyHostname() << "\r\n" <<
+         "ErrPage: " << errorPageName(type) << "\r\n";
+
+    if (xerrno)
+        body << "Err: (" << xerrno << ") " << strerror(xerrno) << "\r\n";
+    else
+        body << "Err: [none]\r\n";
 
-    str.reset();
-    /* email subject line */
-    str.appendf("CacheErrorInfo - %s", errorPageName(type));
-    mb->appendf("?subject=%s", rfc1738_escape_part(str.buf));
-    str.reset();
-    /* email body */
-    str.appendf("CacheHost: %s\r\n", getMyHostname());
-    /* - Err Msgs */
-    str.appendf("ErrPage: %s\r\n", errorPageName(type));
-
-    if (xerrno) {
-        str.appendf("Err: (%d) %s\r\n", xerrno, strerror(xerrno));
-    } else {
-        str.append("Err: [none]\r\n", 13);
-    }
 #if USE_AUTH
     if (auth_user_request.getRaw() && auth_user_request->denyMessage())
-        str.appendf("Auth ErrMsg: %s\r\n", auth_user_request->denyMessage());
+        body << "Auth ErrMsg: " << auth_user_request->denyMessage() << "\r\n";
 #endif
+
     if (dnsError)
-        str.appendf("DNS ErrMsg: " SQUIDSBUFPH "\r\n", SQUIDSBUFPRINT(*dnsError));
+        body << "DNS ErrMsg: " << *dnsError << "\r\n";
 
-    /* - TimeStamp */
-    str.appendf("TimeStamp: %s\r\n\r\n", Time::FormatRfc1123(squid_curtime));
+    body << "TimeStamp: " << Time::FormatRfc1123(squid_curtime) <<
+        "\r\n\r\n";
 
-    /* - IP stuff */
-    str.appendf("ClientIP: %s\r\n", src_addr.toStr(ntoabuf,MAX_IPSTRLEN));
+    char ntoabuf[MAX_IPSTRLEN];
+    body << "ClientIP: " << src_addr.toStr(ntoabuf, MAX_IPSTRLEN) << "\r\n";
 
-    if (request && request->hier.host[0] != '\0') {
-        str.appendf("ServerIP: %s\r\n", request->hier.host);
-    }
+    if (request && request->hier.host[0] != '\0')
+        body << "ServerIP: " << request->hier.host << "\r\n";
+    body << "\r\n";
+
+    body << "HTTP Request:\r\n";
 
-    str.append("\r\n", 2);
-    /* - HTTP stuff */
-    str.append("HTTP Request:\r\n", 15);
     if (request) {
-        str.appendf(SQUIDSBUFPH " " SQUIDSBUFPH " %s/%d.%d\n",
-                    SQUIDSBUFPRINT(request->method.image()),
-                    SQUIDSBUFPRINT(request->url.path()),
-                    AnyP::ProtocolType_str[request->http_ver.protocol],
-                    request->http_ver.major, request->http_ver.minor);
-        request->header.packInto(&str);
+            MemBuf r;
+            r.init();
+            request->pack(&r);
+            body << r.content();
     }
 
-    str.append("\r\n", 2);
```

Comment:
If we are removing this new line after the HTTP Request "section" (when there is no request), should we also move `HTTP Request:` section header inside the `if (request) {}` statement? In other words, if we are changing format (in some cases), then let's improve output (at least in those changed cases).

Reply:
My "output either some information or nothing at all (for every section/item/etc.)" suggestion seems to be completely orthogonal to your "reorder sections" suggestion so I am not sure why you say "rather" (and why you are posting your suggestion in response to mine).

Now that the output diff is clean (thank you for posting that!), I am not against reordering sections if you see some value in doing so.

Reply:
These two (orthogonal) change ideas are still open AFAICT.

Reply:
Should be done now.
I've shuffled some data around. Sample output:

```
CacheHost: n2
ErrPage: ERR_FTP_NOT_FOUND
TimeStamp: Thu, 14 Mar 2024 06:13:04 GMT

ClientIP: 127.0.0.1
ServerIP: www.squid-cache.org

HTTP Request:
GET %2f/foobar HTTP/1.1
Host: www.squid-cache.org
User-Agent: curl/7.81.0
Accept: */*
Proxy-Connection: Keep-Alive

FTP Request: RETR foobar

FTP Reply: Failed to open file.
FTP Msg: 550 Failed to open file.
```

Reply:
```
GET %2f/foobar HTTP/1.1
```

That `%2f` is not really there (in the decoded variant), right?

Reply:
It is, as is in master (and to my knowledge has been in Squid since 2 or even earlier).
See [URL with FTP %2f hack](https://wiki.squid-cache.org/Features/CustomErrors)